### PR TITLE
Add procedure dispatch module

### PR DESCRIPTION
**Pull request by PMunch**
This adds a module which allows very easy dispatch of procedures from
the parsed arguments.

### CODE REVIEW

**Comment by Vindaar** [comment]
I think this would be a really nice addition! I've often wanted to have more type safe control over docopt.

**Comment by PMunch** [comment]
@Vindaar, unfortunately this version of docopt does the parsing of the string and such on runtime. This means that it is not possible to create a type-safe dispatcher without essentially re-implementing docopt. What this does is simply to implement a dispatcher over the table of values that docopt provides. It will throw runtime errors if it is unable to convert the value to the correct type though.

**Comment by Vindaar** [comment]
Oh, yes. I'm aware of that @PMunch. That's why I was talking on IRC the other day about completely changing the docopt "input" structure to either not completely rely on strings, but rather a Nim DSL (imo the string approach is mainly because most languages don't provide Nim's macro powers). Of course then it's not completely docopt anymore. Or alternatively parse the string at CT.

In this context regarding type safety I was only referring to not having to extract the typed data manually from the variant values at runtime.

**Comment by PMunch** [comment]
Ah, yes in that case this is exactly what you want. But yeah, best would be to parse the docopt format on compile-time, then allow you to register dispatch commands and such which would add type-checking and good fallbacks. This would not only make it possible to craft even more type-safe interfaces, but it would also mean that the argument parser would be much faster as it wouldn't have to parse the docopt format every time the program is run.

**Comment by oprypin** [comment]
Regarding this PR: the idea is probably good but I will not reasonably be able to review it. At this point basically any member of Nim community would be a better reviewer than me. I would encourage you to "crowd-source" a (thorough) code review, then I'd take that into account.

**Comment by oprypin** [comment]
Regarding compile-time. Sure, that should be the end-goal and would be an amazing showcase of the strength of the language. Back when I still cared, the blocker was the unavailability of regex at compile time. Now as I understand, that's available. The space is totally open to write a completely new version. (I will not do it, to be clear).
FWIW I do think that it still should be done by parsing the docopt string, definitely not a DSL.

File: src/docopt/dispatch.nim
Comment (oprypin) [inline src/docopt/dispatch.nim:11]: What happens else?

File: docopt.nimble
Comment (oprypin) [inline docopt.nimble:1]: I will not accept a version change in a PR

File: src/docopt/dispatch.nim
Comment (oprypin) [inline src/docopt/dispatch.nim:59]: Wait, is `echo` part of normal operation or is this debugging?

File: examples/dispatch.nim
Comment (oprypin) [inline examples/dispatch.nim:61]: Writing the example like this calls all of the dispatches always, even if one of them matched. It also doesn't preclude the possibility that more than one of them matched. Is that good? Wouldn't you be better served by a short-circuiting `or` operator?

File: src/docopt/dispatch.nim
Comment (Vindaar) [reply src/docopt/dispatch.nim:11]: I think the idea is that these are the only types covered by the `Value` variant object.

I agree though that either `else` should be supplied and given a `{.error: "Invalid type ...".} kind of message or the generic should be restricted to the possible types. The latter is cleaner, imo.

File: src/docopt/dispatch.nim
Comment (Vindaar) [reply src/docopt/dispatch.nim:59]: No, that's debugging output.

File: examples/dispatch.nim
Comment (Vindaar) [reply examples/dispatch.nim:61]: For an example this seems fine to me. As the message in the next line clearly says the intent is simply to check whether anything ran in the first place. 
In practice a user will surely decide how to handle dispatching the procs.

**Comment by PMunch** [comment]
> Regarding compile-time. Sure, that should be the end-goal and would be an amazing showcase of the strength of the language. Back when I still cared, the blocker was the unavailability of regex at compile time. Now as I understand, that's available. The space is totally open to write a completely new version. (I will not do it, to be clear).
> FWIW I do think that it still should be done by parsing the docopt string, definitely not a DSL.

I agree that it should be done by parsing the docopt string and not a DSL. And I'm currently looking into doing a compile-time implementation, but this will be pushed as a separate project.

File: src/docopt/dispatch.nim
Comment (PMunch) [reply src/docopt/dispatch.nim:11]: Definitely should be an error there. I will add one.

File: src/docopt/dispatch.nim
Comment (PMunch) [reply src/docopt/dispatch.nim:59]: Yeah, that just prints out during compile-time what it turned the `dispatchProc` into. But with `--expandMacro` now being implemented in the compiler this should probably be removed.

File: docopt.nimble
Comment (PMunch) [reply docopt.nimble:1]: Why not? This is a new feature after all.

File: examples/dispatch.nim
Comment (PMunch) [reply examples/dispatch.nim:61]: Because of the way docopt works it shouldn't be possible to get more than one of those to match, so only one of them would run. But you're right that doing it this way would conceptually mean that all of the matching procedures would be run, which may or may not be what you wanted to happen.

File: docopt.nimble
Comment (oprypin) [reply docopt.nimble:1]: In my projects (and probably many others) a version change is always done in a commit of itself and by the maintainer.
Not to mention the decision what the actual version should be.

**Comment by oprypin** [comment]
> I would encourage you to "crowd-source" a (thorough) code review, then I'd take that into account.

btw, this is still the case, despite my (minimal) review comments

File: docopt.nimble
Comment (PMunch) [reply docopt.nimble:1]: Ah I see, I'll revert that then

**Comment by PMunch** [comment]
Fixed the various things that were commented about, and added support for custom types in signatures while I was at it

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:5]: Already discussed on IRC, but for completeness: This currently creates code for a call to a `to<CustomType>` proc, even if that doesn't exist. As we agreed it should really emit a `when declared(to<CustomType>)` as a guard.

edit: and mention that this is the magic macro that performs this in a short docstring.

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:28]: Add a docstring explaining why this proc is needed / what it does.

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:30]: Outline the idea of what kind of code will be generated by looking at the arguments.

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:40]: Use `name` and / or at least `.strVal` here?

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:49]: I'd extract `[3]` for clarity in its own variable.

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:51]: Clarify why up to `^3` only.

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:55]: `strVal` (`toStrVal`) again?

**Review by Vindaar — CHANGES_REQUESTED** [review]
Aside from the (already fixed I think) `runUserImplementable` macro, this is mostly nitpicking. Might help other readers who are not that familiar with macros though, which seems important given that this is not really a library depending on macros a lot.

Aside from the mentioned comments, I'd try to give a clearer indication of the ideas of the macro. Similar to what mratsim usually does, e.g. here:

https://github.com/numforge/loop-fusion/blob/master/loopfusion.nim#L95-L114

(I picked this really old example, cause it's the first case in my head about what I mean) A simple empty line + a 1 line comment about what part of the result will be constructed next really helps (even oneself after looking at a macro months later) to quickly understand what each part does.

Again, nitpicking. This is great work!

File: src/docopt/dispatch.nim
Comment (PMunch) [reply src/docopt/dispatch.nim]: This has already been fixed

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:4]: First "checks" is redundant (probably changed your mind mid sentence what to write.

File: src/docopt/dispatch.nim
Comment (Vindaar) [inline src/docopt/dispatch.nim:42]: typo ".o" at end of sentence.

**Comment by Vindaar** [comment]
Looks really great to me now (barring the 2 small typos)!

**Comment by PMunch** [comment]
Rebased onto the current master and fixed compilation on Nim 1.4.0.

**Comment by oprypin** [comment]
Can anyone vouch that 2fdff42518c891422609edcbb14b800d6c41cfe1 is the simplest solution?

**Comment by oprypin** [comment]
[Reference](https://nim-lang.org/blog/2020/10/16/version-140-released.html)

> `getImpl` on enum type symbols now returns field syms instead of idents. This helps with writing typed macros. The old behavior for backwards compatibility can be restored with `--useVersion:1.0`.

**Comment by PMunch** [comment]
It seems to be pretty much the only way to do it: https://irclogs.nim-lang.org/02-11-2020.html#11:25:51, and from the creator of Nim himself: https://irclogs.nim-lang.org/02-11-2020.html#12:47:14. So at least until a new version of Nim comes around with something like this in the standard library this is the easiest (and pretty much only) way. I'll make a PR to swap it out if something like this ends up in the stdlib though.

As for the thing you copy-pasted, forcing people to use the `--useVersion:1.0` isn't great as it has project-wide consequences. And creating a solution like this that works both on older version and is the way going forward is decidedly a better option.

**Comment by oprypin** [comment]
Oh I wasn't implying anything about `--useVersion:1.0`.

OK thanks. I'll probably merge this after I take another look.

**Comment by Vindaar** [comment]
> Can anyone vouch that [2fdff42](https://github.com/docopt/docopt.nim/commit/2fdff42518c891422609edcbb14b800d6c41cfe1) is the simplest solution?

Aside from what PMunch already answered, yes it really is. I use similar code in my macros where required and other people do the same.

**Comment by PMunch** [comment]
Still no word on this?

**Comment by oprypin** [comment]
Hi.
Yeaaah at this point I can say with certainty: 
This project is abandoned.
Sorry.

**Comment by oprypin** [comment]
Or actually at that point, why not just merge it :D
I'll do that today and release afterwards

**Comment by grey4owl** [comment]
@oprypin Hey, can you just release new version because right now if you install via `nimble install docopt` the dispatch module doesn't work! Actually if you go to the directorium `~/.nimble/pkgs/docopt-0.6.8/docopt/` there is no `dispatch.nim` file..

Thanks :relaxed:

**Comment by oprypin** [comment]
Aah sorry :cry: 
I've been looking at it in my todo list for a while and not actually doing it

**Comment by oprypin** [comment]
I tagged https://github.com/docopt/docopt.nim/releases/tag/v0.7.0 now

**Comment by grey4owl** [comment]
@oprypin And dispatch module still doesn't work, it gives me the following error on compile: `Error: attempting to call undeclared routine: 'strip'` in `dispatch.nim (ln 99, 19)`?

**Comment by oprypin** [comment]
Well, I have no idea how Nim works by now, contributions definitely welcome.

**Comment by PMunch** [comment]
@grey4owl I'll have a look at this after the weekend. For now you probably have to use an older Nim version

**Comment by grey4owl** [comment]
@PMunch I partially fixed it, check out https://github.com/docopt/docopt.nim/pull/54

**Comment by PMunch** [comment]
@grey4owl, not entirely sure how you ended up with that error. If I clone the repo, `cd` into the examples folder, and compile `dispatch.nim` it works just fine. And I'm able to run it and get it to dispatch as I would expect. This is with both latest Nim and devel.

**Comment by oprypin** [comment]
And everything passes in CI, including the dispatch example
https://github.com/docopt/docopt.nim/actions/runs/3210603085
https://github.com/docopt/docopt.nim/blob/master/examples/dispatch.nim
https://github.com/docopt/docopt.nim/blob/master/docopt.nimble

**Comment by grey4owl** [comment]
@PMunch Everything works fine now.. I wonder if it is possible to set a custom message when a wrong / non-existent argument is entered and when no argument is entered so that it does not show the default usage message every time but something custom? I already made a custom help menu, I just set `help = false` and made my own, but I would also like to change those messages so that they don't show the default usage string from the doc..

**Comment by grey4owl** [comment]
@PMunch Actually I made a custom message if no argument was entered, I'm only interested in this one with the wrong / non-existent argument, but how do I handle the output message if a wrong / non-existent argument is entered?

**Comment by grey4owl** [comment]
@PMunch Actually I made a custom message if no argument was entered, but how do I handle the output message if a wrong / non-existent argument is entered?

**Pull request merged**
